File: src/client/java/fr/hardel/asset_editor/client/javafx/routes/enchantment/EnchantmentTechnicalPage.java
package fr.hardel.asset_editor.client.javafx.routes.enchantment;

import fr.hardel.asset_editor.client.javafx.VoxelColors;
import fr.hardel.asset_editor.client.javafx.VoxelFonts;
import fr.hardel.asset_editor.client.javafx.components.ui.Range;
import fr.hardel.asset_editor.client.javafx.components.ui.ResponsiveGrid;
import fr.hardel.asset_editor.client.javafx.components.ui.Section;
import fr.hardel.asset_editor.client.javafx.components.ui.SwitchCard;
import fr.hardel.asset_editor.client.javafx.lib.RegistryPage;
import fr.hardel.asset_editor.client.javafx.lib.StudioContext;
import fr.hardel.asset_editor.client.javafx.lib.action.EnchantmentActions;
import fr.hardel.asset_editor.client.javafx.lib.data.StudioBreakpoint;
import fr.hardel.asset_editor.client.javafx.lib.StudioText;
import fr.hardel.asset_editor.store.ElementEntry;
import fr.hardel.asset_editor.network.EditorAction;
import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.scene.control.Label;
import javafx.scene.layout.VBox;
import net.minecraft.client.resources.language.I18n;
import net.minecraft.core.registries.Registries;
import net.minecraft.resources.Identifier;
import net.minecraft.world.item.enchantment.Enchantment;

import java.util.List;
import java.util.function.Function;
import java.util.function.IntFunction;
import java.util.function.UnaryOperator;

public final class EnchantmentTechnicalPage extends RegistryPage<Enchantment> {

    private record BehaviourTag(Identifier tagId) {
        String titleKey() { return "enchantment_tag:" + tagId; }
        String descKey() { return "enchantment_tag:" + tagId + ".desc"; }
    }

    private record CostDef(String key,
            Function<ElementEntry<Enchantment>, Integer> selector,
            IntFunction<UnaryOperator<Enchantment>> mutation,
            IntFunction<EditorAction> actionFactory) {
    }

    private static final List<BehaviourTag> BEHAVIOUR_TAGS = List.of(
            new BehaviourTag(EnchantmentActions.CURSE_TAG),
            new BehaviourTag(EnchantmentActions.DOUBLE_TRADE_PRICE_TAG),
            new BehaviourTag(EnchantmentActions.PREVENTS_BEE_SPAWNS_WHEN_MINING_TAG),
            new BehaviourTag(EnchantmentActions.PREVENTS_DECORATED_POT_SHATTERING_TAG),
            new BehaviourTag(EnchantmentActions.PREVENTS_ICE_MELTING_TAG),
            new BehaviourTag(EnchantmentActions.PREVENTS_INFESTED_SPAWNS_TAG),
            new BehaviourTag(EnchantmentActions.SMELTS_LOOT_TAG));

    private static final List<CostDef> COST_FIELDS = List.of(
            new CostDef("minCostBase",
                    entry -> entry.data().definition().minCost().base(),
                    EnchantmentActions::minCostBase,
                    v -> new EditorAction.SetIntField("min_cost_base", v)),
            new CostDef("minCostPerLevelAboveFirst",
                    entry -> entry.data().definition().minCost().perLevelAboveFirst(),
                    EnchantmentActions::minCostPerLevel,
                    v -> new EditorAction.SetIntField("min_cost_per_level", v)),
            new CostDef("maxCostBase",
                    entry -> entry.data().definition().maxCost().base(),
                    EnchantmentActions::maxCostBase,
                    v -> new EditorAction.SetIntField("max_cost_base", v)),
            new CostDef("maxCostPerLevelAboveFirst",
                    entry -> entry.data().definition().maxCost().perLevelAboveFirst(),
                    EnchantmentActions::maxCostPerLevel,
                    v -> new EditorAction.SetIntField("max_cost_per_level", v)));

    public EnchantmentTechnicalPage(StudioContext context) {
        super(context, Registries.ENCHANTMENT, "enchantment-subpage-scroll", 64, new Insets(32));
    }

    @Override
    protected void buildContent() {
        Section behaviour = new Section(I18n.get("enchantment:section.technical.description"));
        ResponsiveGrid tagGrid = new ResponsiveGrid(ResponsiveGrid.fixed(1))
                .atLeast(StudioBreakpoint.LG, ResponsiveGrid.fixed(1, 1));

        for (BehaviourTag field : BEHAVIOUR_TAGS) {
            SwitchCard sw = SwitchCard.literal(I18n.get(field.titleKey()), I18n.get(field.descKey()));
            bindToggle(sw.valueProperty(), entry -> entry.tags().contains(field.tagId()),
                    () -> applyTagToggle(field.tagId()));
            tagGrid.addItem(sw);
        }
        behaviour.addContent(tagGrid);

        Section costs = new Section(I18n.get("enchantment:section.costs"));
        ResponsiveGrid costGrid = new ResponsiveGrid(ResponsiveGrid.fixed(1))
                .atLeast(StudioBreakpoint.LG, ResponsiveGrid.fixed(1, 1));

        for (CostDef field : COST_FIELDS) {
            Range range = new Range(I18n.get("enchantment:global." + field.key() + ".title"), 0, 100, 1, 0);
            bindInt(range.valueProperty(), field.selector(), field.mutation(), field.actionFactory());
            costGrid.addItem(range);
        }
        costs.addContent(costGrid);

        content().getChildren().setAll(behaviour, costs, buildEffectsSection());
    }

    private VBox buildEffectsSection() {
        Section section = new Section(I18n.get("enchantment:technical.effects.title"));
        var effectsSelector = select(entry -> EnchantmentActions.availableEffects(entry.data()));
        List<String> effects = effectsSelector.get() == null ? List.of() : effectsSelector.get();

        if (effects.isEmpty()) {
            Label empty = new Label(I18n.get("enchantment:technical.empty_effects"));
            empty.setFont(VoxelFonts.of(VoxelFonts.Variant.REGULAR, 14));
            empty.setTextFill(VoxelColors.ZINC_400);
            empty.setPadding(new Insets(16, 0, 16, 0));
            empty.setMaxWidth(Double.MAX_VALUE);
            empty.setAlignment(Pos.CENTER);
            section.addContent(empty);
            return section;
        }

        ResponsiveGrid grid = new ResponsiveGrid(ResponsiveGrid.fixed(1))
                .atLeast(StudioBreakpoint.LG, ResponsiveGrid.fixed(1, 1));

        for (String effectId : effects) {
            Identifier effectKey = Identifier.tryParse(effectId);
            String effectLabel = effectKey != null
                    ? StudioText.resolve("effect", effectKey)
                    : effectId;
            String descKey = effectKey != null
                    ? "effect:" + effectKey + ".desc"
                    : "";
            String effectDesc = I18n.exists(descKey) ? I18n.get(descKey) : effectId;
            SwitchCard card = SwitchCard.literal(effectLabel, effectDesc);
            bindToggle(card.valueProperty(),
                    entry -> !EnchantmentActions.isEffectDisabled(entry, effectId),
                    () -> applyCustomAction(EnchantmentActions.toggleDisabledEffect(effectId),
                            new EditorAction.ToggleDisabledEffect(effectId)));
            grid.addItem(card);
        }

        section.addContent(grid);
        return section;
    }
}

File: src/client/kotlin/fr/hardel/asset_editor/client/compose/routes/enchantment/EnchantmentTechnicalPage.kt
package fr.hardel.asset_editor.client.compose.routes.enchantment

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import fr.hardel.asset_editor.client.compose.StudioBreakpoint
import fr.hardel.asset_editor.client.compose.VoxelColors
import fr.hardel.asset_editor.client.compose.VoxelTypography
import fr.hardel.asset_editor.client.compose.components.ui.*
import fr.hardel.asset_editor.client.compose.lib.*
import fr.hardel.asset_editor.workspace.flush.adapter.EnchantmentFlushAdapter
import fr.hardel.asset_editor.workspace.flush.Workspaces
import fr.hardel.asset_editor.workspace.action.EditorAction
import fr.hardel.asset_editor.workspace.action.enchantment.SetIntFieldAction
import fr.hardel.asset_editor.workspace.action.enchantment.ToggleDisabledEffectAction
import fr.hardel.asset_editor.workspace.action.enchantment.ToggleTagAction
import net.minecraft.client.resources.language.I18n
import net.minecraft.core.registries.Registries
import net.minecraft.world.item.enchantment.Enchantment
import net.minecraft.resources.Identifier

private data class BehaviourTag(val tagId: Identifier) {
    fun titleKey(): String = "enchantment_tag:$tagId"
    fun descKey(): String = "enchantment_tag:$tagId.desc"
}

private data class CostField(val key: String, val value: (Enchantment) -> Int, val action: (Int) -> EditorAction<*>)

private val BEHAVIOUR_TAGS = listOf(
    BehaviourTag(EnchantmentFlushAdapter.SMELTS_LOOT_TAG),
    BehaviourTag(EnchantmentFlushAdapter.PREVENTS_ICE_MELTING_TAG),
    BehaviourTag(EnchantmentFlushAdapter.PREVENTS_INFESTED_SPAWNS_TAG),
    BehaviourTag(EnchantmentFlushAdapter.PREVENTS_BEE_SPAWNS_WHEN_MINING_TAG),
    BehaviourTag(EnchantmentFlushAdapter.PREVENTS_DECORATED_POT_SHATTERING_TAG),
    BehaviourTag(EnchantmentFlushAdapter.DOUBLE_TRADE_PRICE_TAG)
)

private val COST_FIELDS = listOf(
    CostField(
        "minCostBase",
        { enchantment -> enchantment.definition().minCost().base() },
        { value -> SetIntFieldAction("min_cost_base", value) }),
    CostField(
        "minCostPerLevelAboveFirst",
        { enchantment -> enchantment.definition().minCost().perLevelAboveFirst() },
        { value -> SetIntFieldAction("min_cost_per_level", value) }),
    CostField(
        "maxCostBase",
        { enchantment -> enchantment.definition().maxCost().base() },
        { value -> SetIntFieldAction("max_cost_base", value) }),
    CostField(
        "maxCostPerLevelAboveFirst",
        { enchantment -> enchantment.definition().maxCost().perLevelAboveFirst() },
        { value -> SetIntFieldAction("max_cost_per_level", value) })
)

@Composable
fun EnchantmentTechnicalPage(context: StudioContext) {
    val dialogs = rememberRegistryDialogState()
    val entry = rememberCurrentRegistryEntry(context, Registries.ENCHANTMENT) ?: return
    val effects = remember(entry) { EnchantmentFlushAdapter.availableEffects(entry.data()) }

    Column(
        verticalArrangement = Arrangement.spacedBy(32.dp),
        modifier = Modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState())
            .padding(32.dp)
    ) {
        Section(I18n.get("enchantment:section.technical.description")) {
            ResponsiveGrid(
                items = BEHAVIOUR_TAGS.map { field ->
                    {
                        SwitchCard(
                            title = I18n.get(field.titleKey()),
                            description = I18n.get(field.descKey()),
                            checked = entry.tags().contains(field.tagId),
                            onCheckedChange = {
                                context.dispatchRegistryAction(
                                    definition = Workspaces.ENCHANTMENT,
                                    target = entry.id(),
                                    action = ToggleTagAction(field.tagId),
                                    dialogs = dialogs
                                )
                            }
                        )
                    }
                },
                defaultSpec = LayoutSpec.Fixed(floatArrayOf(1f)),
                rules = listOf(
                    BreakpointRule(
                        minWidth = StudioBreakpoint.LG.px.dp,
                        spec = LayoutSpec.Fixed(floatArrayOf(1f, 1f))
                    )
                )
            )
        }

        Section(I18n.get("enchantment:section.costs")) {
            ResponsiveGrid(
                items = COST_FIELDS.map { field ->
                    {
                        Range(
                            label = I18n.get("enchantment:global.${field.key}.title"),
                            value = field.value(entry.data()),
                            onValueChange = { value ->
                                context.dispatchRegistryAction(
                                    definition = Workspaces.ENCHANTMENT,
                                    target = entry.id(),
                                    action = field.action(value),
                                    dialogs = dialogs
                                )
                            },
                            min = 0,
                            max = 100,
                            step = 1
                        )
                    }
                },
                defaultSpec = LayoutSpec.Fixed(floatArrayOf(1f)),
                rules = listOf(
                    BreakpointRule(
                        minWidth = StudioBreakpoint.LG.px.dp,
                        spec = LayoutSpec.Fixed(floatArrayOf(1f, 1f))
                    )
                )
            )
        }

        Section(I18n.get("enchantment:technical.effects.title")) {
            if (effects.isEmpty()) {
                Text(
                    text = I18n.get("enchantment:technical.empty_effects"),
                    style = VoxelTypography.regular(14),
                    color = VoxelColors.Zinc400,
                    modifier = Modifier.padding(vertical = 16.dp)
                )
            } else {
                ResponsiveGrid(
                    items = effects.map { effectId ->
                        {
                            val effectKey = Identifier.tryParse(effectId)
                            val effectLabel =
                                if (effectKey != null) StudioText.resolve("effect", effectKey) else effectId
                            val descKey = if (effectKey != null) "effect:$effectKey.desc" else ""
                            SwitchCard(
                                title = effectLabel,
                                description = if (descKey.isNotBlank() && I18n.exists(descKey)) I18n.get(descKey) else effectId,
                                checked = !EnchantmentFlushAdapter.isEffectDisabled(entry, effectId),
                                onCheckedChange = {
                                    context.dispatchRegistryAction(
                                        definition = Workspaces.ENCHANTMENT,
                                        target = entry.id(),
                                        action = ToggleDisabledEffectAction(effectId),
                                        dialogs = dialogs
                                    )
                                }
                            )
                        }
                    },
                    defaultSpec = LayoutSpec.Fixed(floatArrayOf(1f)),
                    rules = listOf(
                        BreakpointRule(
                            minWidth = StudioBreakpoint.LG.px.dp,
                            spec = LayoutSpec.Fixed(floatArrayOf(1f, 1f))
                        )
                    )
                )
            }
        }
    }

    RegistryPageDialogs(context, dialogs)
}
